src/Microsoft.AspNet.Hosting/HostingEngine.cs

```diff
@@ -5,56 +5,73 @@
+        private readonly IServiceProvider _fallbackServices;
+        private readonly ApplicationLifetime _appLifetime;
+        private IServerLoader _serverLoader;
+        private IApplicationBuilderFactory _builderFactory;
+
+        // Move everything except startupLoader to get them after configureServices
+        public HostingEngine() : this(fallbackServices: null) { }
+
+        public HostingEngine(IServiceProvider fallbackServices)
         {
-            _serverManager = serverManager;
-            _startupLoader = startupLoader;
-            _builderFactory = builderFactory;
-            _httpContextFactory = httpContextFactory;
-            _contextAccessor = contextAccessor;
+            _fallbackServices = fallbackServices ?? CallContextServiceLocator.Locator.ServiceProvider;
+            _appLifetime = new ApplicationLifetime();
         }
 
         public IDisposable Start(HostingContext context)
         {
+            EnsureContextDefaults(context);
+            EnsureApplicationServices(context);
             EnsureBuilder(context);
             EnsureServerFactory(context);
             InitalizeServerFactory(context);
             EnsureApplicationDelegate(context);
 
-            var applicationLifetime = (ApplicationLifetime)context.ApplicationLifetime;
-            var pipeline = new PipelineInstance(_httpContextFactory, context.ApplicationDelegate, _contextAccessor);
-            var server = context.ServerFactory.Start(context.Server, pipeline.Invoke);
+            var contextFactory = context.ApplicationServices.GetRequiredService<IHttpContextFactory>();
+            var contextAccessor = context.ApplicationServices.GetRequiredService<IHttpContextAccessor>();
+            var server = context.ServerFactory.Start(context.Server, 
+                features =>
+                {
+                    var httpContext = contextFactory.CreateHttpContext(features);
+                    contextAccessor.HttpContext = httpContext;
+                    return context.ApplicationDelegate(httpContext);
+                });
 
             return new Disposable(() =>
             {
-                applicationLifetime.NotifyStopping();
+                _appLifetime.NotifyStopping();
                 server.Dispose();
-                pipeline.Dispose();
-                applicationLifetime.NotifyStopped();
+                _appLifetime.NotifyStopped();
             });
         }
 
+        private void EnsureContextDefaults(HostingContext context)
+        {
+            var appEnv = _fallbackServices.GetRequiredService<IApplicationEnvironment>();
+            if (context.ApplicationName == null)
+            {
+                context.ApplicationName = appEnv.ApplicationName;
+            }
+            if (context.EnvironmentName == null)
+            {
+                context.EnvironmentName = context.Configuration?.Get(EnvironmentKey) ?? HostingEnvironment.DefaultEnvironmentName;
```

Comment:
It should come as part of CreateHostingServices, the defaults get configured before that step, but the service should be a singleton in the builder.ApplicationServices